pkg/globalnet/controllers/ingress_endpoints_controller.go

```diff
@@ -0,0 +1,203 @@
+		ingressIPs:           config.SourceClient.Resource(*gvr),
+	}
+
+	fieldSelector := fields.Set(map[string]string{"metadata.name": svc.Name}).AsSelector().String()
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Endpoints syncer",
+		ResourceType:        &corev1.Endpoints{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     svc.Namespace,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		SourceFieldSelector: fieldSelector,
+		ResourcesEquivalent: areEndpointsEqual,
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating the syncer")
+	}
+
+	if err := controller.Start(); err != nil {
+		return nil, errors.Wrap(err, "error creating the syncer")
+	}
+
+	ingressIPs := config.SourceClient.Resource(*gvr).Namespace(corev1.NamespaceAll)
+	controller.reconcile(ingressIPs, "" /* labelSelector */, fieldSelector, func(obj *unstructured.Unstructured) runtime.Object {
+		endpointsName, exists, _ := unstructured.NestedString(obj.Object, "spec", "serviceRef", "name")
+		if exists {
+			return &corev1.Endpoints{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      endpointsName,
+					Namespace: obj.GetNamespace(),
+				},
+			}
+		}
+
+		return nil
+	})
+
+	klog.Infof("Created Endpoints controller for (%s/%s) with selector %q", svc.Namespace, svc.Name, fieldSelector)
+
+	return controller, nil
+}
+
+func (c *ingressEndpointsController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	endpoints := from.(*corev1.Endpoints)
+
+	switch op {
+	case syncer.Create, syncer.Update:
+		return c.onCreateOrUpdate(endpoints, op)
+	case syncer.Delete:
+		return c.onDelete(endpoints)
+	}
+
+	return nil, false
+}
+
+func (c *ingressEndpointsController) onCreateOrUpdate(endpoints *corev1.Endpoints, op syncer.Operation) (runtime.Object, bool) {
+	key, _ := cache.MetaNamespaceKeyFunc(endpoints)
+
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// Get list of current endpoint IPs
+	epIPs := map[string]bool{}
+
+	for _, subset := range endpoints.Subsets {
+		for _, addr := range subset.Addresses {
+			epIPs[addr.IP] = true
+		}
+	}
+
+	// Get List of existing ingressIPs
+	selector := labels.SelectorFromSet(map[string]string{ServiceRefLabel: endpoints.Name}).String()
+
+	existingIngressIPs, err := c.ingressIPs.Namespace(endpoints.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return nil, true
+	}
+
+	// Delete unused ingressIPs
+	for _, ingressIP := range existingIngressIPs.Items {
+		annotations := ingressIP.GetAnnotations()
+		if epIP, ok := annotations[headlessSvcEndpointsIP]; ok {
+			if _, used := epIPs[epIP]; used {
+				// ingressIP is still used
```

Comment:
If the ingressIP is still used, should we remove it from the epIP list? As otherwise we seem to create new IngressIPs in Line 152. Am I missing something?